libs/platform/ebpf_object.c

```diff
@@ -183,64 +183,45 @@ ebpf_object_initialize(
 }
 
 void
-ebpf_object_acquire_reference(ebpf_core_object_t* object)
+ebpf_object_acquire_reference(_Inout_ ebpf_core_object_t* object)
 {
     ebpf_assert(object->base.marker == _ebpf_object_marker);
     int32_t new_ref_count = ebpf_interlocked_increment_int32(&object->base.reference_count);
     ebpf_assert(new_ref_count != 1);
 }
 
-_Requires_lock_held_(&_ebpf_object_tracking_list_lock) static void _ebpf_object_release_reference_under_lock(
-    ebpf_core_object_t* object)
+void
+ebpf_object_release_reference(_Inout_opt_ ebpf_core_object_t* object)
 {
     int32_t new_ref_count;
 
-    if (!object) {
+    if (!object)
         return;
-    }
 
+    ebpf_lock_state_t state = ebpf_lock_lock(&_ebpf_object_tracking_list_lock);
     ebpf_assert(object->base.marker == _ebpf_object_marker);
 
     new_ref_count = ebpf_interlocked_decrement_int32(&object->base.reference_count);
     ebpf_assert(new_ref_count != -1);
 
+    // Remove from object tracking list under the lock.
     if (new_ref_count == 0) {
         EBPF_LOG_MESSAGE_POINTER_ENUM(
             EBPF_TRACELOG_LEVEL_VERBOSE, EBPF_TRACELOG_KEYWORD_BASE, "eBPF object terminated", object, object->type);
-
         _ebpf_object_tracking_list_remove(object);
-        object->base.marker = ~object->base.marker;
-        object->free_function(object);
-    }
-}
-
-void
-ebpf_object_release_reference(ebpf_core_object_t* object)
-{
-    int32_t new_ref_count;
-
-    if (!object) {
-        return;
     }
 
-    ebpf_assert(object->base.marker == _ebpf_object_marker);
-
-    new_ref_count = ebpf_interlocked_decrement_int32(&object->base.reference_count);
-    ebpf_assert(new_ref_count != -1);
+    ebpf_lock_unlock(&_ebpf_object_tracking_list_lock, state);
```

Comment:
question: is this in the hot path and will this adversely affect perf?

Reply:
Yes, ebpf_program_invoke calls ebpf_object_release_reference for tail calls. Let me see if I can figure out a way to do this without always taking the lock.

Reply:
Note: A better fix would be to remove the entire interlocked operation from the invoke path.

Reply:
https://github.com/microsoft/ebpf-for-windows/issues/2314

The code is already contending on the _ebpf_object_tracking_list_lock lock during tail calls and this change merely increases the number of times the lock is acquired. In either case, it won't scale well and needs to be fixed.